### ISSUE TITLE
Explore the possibility of using Tangent for gradients in Poincare embeddings

### ISSUE THREAD

**Issue by jayantj** [RaRe-Technologies/gensim#1778]
The current implementation makes use of [autograd](https://github.com/HIPS/autograd/) to auto-differentiate the loss function. As the automatic differentiation is too slow, we chose to not use this feature for training, instead computing the derivatives ourselves and adding an option to perform verification of the computed derivatives by comparing it to the autograd values every 10k iterations or so. This is a very useful feature for debugging.

Another library to consider is [Tangent](https://github.com/google/tangent) library (thanks to @alexbw for the suggestion), which does ahead-of-time gradient computations. The main concern here is speed - auto-differentiation should be similar to or faster than computing the derivatives ourselves.

**Comment by menshikh-iv** [comment]
@jayantj can you suggest some way "how to test it" in a better way? I played with Tangent, it works quite well, but that is not enough.

**Comment by alexbw** [comment]
Tangent should hopefully be a drop-in replacement for autograd. However,
tangent is new, and might require some fixes to fully work here. Please let
me know what issues you run into and I'd be happy to help.

**Comment by menshikh-iv** [comment]
@alexbw thanks, we will definitely let you know if we find any problems
